### PR TITLE
Improve metadata tests (add CBOR parser)

### PR DESCRIPTION
**Pull request by axic**
Prerequisite for #4764.

### CODE REVIEW

File: test/Metadata.cpp [hunk 1]
```diff
@@ -51,6 +61,99 @@ string bytecodeSansMetadata(string const& _bytecode)
...
+					return "true";
+				else
+					assertThrow(false, CBORException, "Unsupported simple value (not a boolean).");
+			}
+			default:
+				assertThrow(false, CBORException, "Unsupported value type.");
+		}
+	}
+private:
+	unsigned nextType() const { return (m_metadata[m_pos] >> 5) & 0x7; }
```
Comment (chriseth) [inline test/Metadata.cpp]: No length check? You could just use `.at` everywhere.

File: test/Metadata.cpp [hunk 2]
```diff
@@ -51,6 +61,99 @@ string bytecodeSansMetadata(string const& _bytecode)
...
 
+DEV_SIMPLE_EXCEPTION(CBORException);
+
+class TinyCBORParser
+{
+public:
+	explicit TinyCBORParser(bytes const& _metadata): m_pos(0), m_metadata(_metadata) {}
+	unsigned mapItemCount()
+	{
+		assertThrow(nextType() == 5, CBORException, "Fixed-length map expected.");
```
Comment (chriseth) [inline test/Metadata.cpp]: Enum for types?

**Comment by axic** [comment]
@chriseth made some updates, will squash it

File: test/Metadata.cpp [hunk 3]
```diff
@@ -66,10 +66,13 @@ DEV_SIMPLE_EXCEPTION(CBORException);
 class TinyCBORParser
 {
 public:
-	explicit TinyCBORParser(bytes const& _metadata): m_pos(0), m_metadata(_metadata) {}
+	explicit TinyCBORParser(bytes const& _metadata): m_pos(0), m_metadata(_metadata)
+	{
+		assertThrow((_metadata.size() - m_pos) > 0, CBORException, "Input too short.");
```
Comment (chriseth) [inline test/Metadata.cpp]: They are both unsigned, so probably `m_pos < _metadata.size()` would be better.

**Comment by chriseth** [comment]
Looks good apart from the one change.

File: test/Metadata.cpp (hunk 3 above)
Comment (axic) [reply test/Metadata.cpp]: Ok! Will change that and squash it into the right commit.

File: test/Metadata.cpp [hunk 4]
```diff
@@ -59,6 +61,116 @@ string bytecodeSansMetadata(string const& _bytecode)
...
+		}
+	}
+	unsigned nextImmediate() const { return m_metadata.at(m_pos) & 0x1f; }
+	unsigned readLength()
+	{
+		unsigned length = m_metadata.at(m_pos++) & 0x1f;
+		if (length < 24)
+			return length;
+		if (length == 24)
+			return m_metadata.at(m_pos++);
```
Comment (axic) [inline test/Metadata.cpp:132]: Could have a length check here assertion here, but the entire class is in a try/catch loop for this reason in the caller.

**Review by chriseth — APPROVED** [review]

**Pull request merged**